code/modules/client/preferences/middleware/species.dm

```diff
@@ -31,3 +31,21 @@
 
 	for (var/spritesheet_key in to_insert)
 		Insert(spritesheet_key, to_insert[spritesheet_key])
+
+// Remove any invalid quirks by the species whitelist
+/datum/preference_middleware/species/post_set_preference(mob/user, preference, value)
+	. = ..()
+	var/species_id = value // one of the options that this can be is species_id, but it can be other things as well annoyingly
+	if(preference != "species" || !length(GLOB.quirk_species_whitelist) || !species_id)
+		return .
+	var/needs_update = FALSE
+	for(var/quirk_name as anything in preferences.all_quirks)
+		var/datum/quirk/quirk_type = SSquirks.quirks[quirk_name]
+		var/list/species_whitelist = GLOB.quirk_species_whitelist[quirk_type]
+		if(!length(species_whitelist) || (species_id in species_whitelist))
+			continue
+		preferences.all_quirks -= quirk_name
```

Comment:
There needs to be a confirmation popup when you click on a species that's going to change your quirks to let you know that it's going to happen (and which one), I've asked this of similar PRs in the past